webapp/survey/nodeDefs/actions.js

```diff
@@ -173,25 +188,106 @@ export const putNodeDefLayoutProp = (nodeDef, key, value) => async (dispatch, ge
       return layout
     }),
   )(nodeDef)
+}
+
+/**
+ * Updates the specified layout prop of the node def being edited, without persisting the change
+ */
+export const setNodeDefLayoutProp = (key, value) => async (dispatch, getState) => {
+  const state = getState()
+  const nodeDef = NodeDefState.getNodeDef(state)
+  const layoutUpdated = _updateLayoutProp(getState, nodeDef, key, value)
+
+  dispatch(setNodeDefProp(NodeDefLayout.keys.layout, layoutUpdated))
+}
+
+/**
+ * Updates the specified layout prop of a node def and persists the change
+ */
+export const putNodeDefLayoutProp = (nodeDef, key, value) => async (dispatch, getState) => {
+  const layoutUpdated = _updateLayoutProp(getState, nodeDef, key, value)
+  const props = { [NodeDefLayout.keys.layout]: layoutUpdated }
+  const nodeDefUpdated = NodeDef.mergeProps(props)(nodeDef)
+  dispatch({ type: nodeDefUpdate, nodeDef: nodeDefUpdated })
+
+  dispatch(_putNodeDefPropsDebounced(nodeDef, NodeDefLayout.keys.layout, props))
+}
+
+export const cancelNodeDefEdits = history => async (dispatch, getState) => {
+  const state = getState()
+  const nodeDef = NodeDefState.getNodeDef(state)
+  const nodeDefOriginal = NodeDefState.getNodeDefOriginal(state)
+  const i18n = AppState.getI18n(state)
 
-  const checkFormPageUuid = R.includes(key, [NodeDefLayout.keys.renderType, NodeDefLayout.keys.pageUuid])
-  dispatch(putNodeDefProp(nodeDef, NodeDefLayout.keys.layout, layoutUpdate, false, checkFormPageUuid))
+  if (!NodeDefState.isDirty(state) || confirm(i18n.t('surveyForm.nodeDefEditFormActions.confirmCancel'))) {
+    dispatch({
+      type: nodeDefPropsTempCancel,
+      nodeDef,
+      nodeDefOriginal,
+      isNodeDefNew: NodeDef.isTemporary(nodeDef),
+    })
+    history.goBack()
+  }
+}
+
+/**
+ * Persists the temporary changes applied to the node def in the state
+ */
+export const saveNodeDefEdits = () => async (dispatch, getState) => {
+  const state = getState()
+  const survey = SurveyState.getSurvey(state)
+  const surveyId = SurveyState.getSurveyId(state)
+  const cycle = SurveyState.getSurveyCycleKey(state)
+  const nodeDef = NodeDefState.getNodeDef(state)
+
+  dispatch(showAppLoader())
+
+  const isNodeDefNew = NodeDef.isTemporary(nodeDef)
+  const nodeDefUpdated = NodeDef.dissocTemporary(nodeDef)
+
+  if (isNodeDefNew) {
+    const {
+      data: { nodeDefsValidation, nodeDefsUpdated },
+    } = await axios.post(`/api/survey/${surveyId}/nodeDef`, nodeDef)
+    _onNodeDefsUpdate(dispatch, nodeDefsUpdated, nodeDefsValidation)
```

Comment:
i would use ```dispatch(_onNodeDefsUpdate(nodeDefsUpdated, nodeDefsValidation))```

Reply:
it's an internal method, but just to make the calls consistent I'll change it.